Sources/Testing/SourceAttribution/Expression.swift

```diff
@@ -0,0 +1,261 @@
+  ///     resulting string in parentheses (as needed depending on what
+  ///     information this instance contains.)
+  ///
+  /// - Returns: A string describing this instance.
+  func expandedDescription(includingParenthesesIfNeeded: Bool = true) -> String {
+    switch kind {
+    case let .generic(sourceCode):
+      let runtimeValueDescription = runtimeValueDescription ?? "<not evaluated>"
+      return if runtimeValueDescription == "(Function)" {
+        // Hack: don't print string representations of function calls.
+        sourceCode
+      } else if runtimeValueDescription == sourceCode {
+        sourceCode
+      } else if includingParenthesesIfNeeded {
+        "(\(sourceCode) → \(runtimeValueDescription))"
+      } else {
+        "\(sourceCode) → \(runtimeValueDescription)"
+      }
+    case let .binaryOperation(lhsExpr, op, rhsExpr):
+      return "\(lhsExpr.expandedDescription()) \(op) \(rhsExpr.expandedDescription())"
+    case let .functionCall(value, functionName, arguments):
+      let includeParentheses = arguments.count > 1
+      let argumentList = arguments.lazy
+        .map { argument in
+          (argument.label, argument.value.expandedDescription(includingParenthesesIfNeeded: includeParentheses))
+        }.map { label, value in
+          if let label {
+            return "\(label): \(value)"
+          }
+          return value
+        }.joined(separator: ", ")
+      if let value {
+        return "\(value.expandedDescription()).\(functionName)(\(argumentList))"
+      }
+      return "\(functionName)(\(argumentList))"
+    case let .propertyAccess(value, keyPath):
+      return "\(value.expandedDescription()).\(keyPath.expandedDescription(includingParenthesesIfNeeded: false))"
+    }
+  }
+
+  /// The set of parsed and captured subexpressions contained in this instance.
+  @_spi(ExperimentalSourceCodeCapturing)
+  public var subexpressions: [Expression] {
+    switch kind {
+    case .generic:
+      []
+    case let .binaryOperation(lhs, _, rhs):
+      [lhs, rhs]
+    case let .functionCall(value, _, arguments):
+      if let value {
+        CollectionOfOne(value) + arguments.lazy.map(\.value)
+      } else {
+        arguments.lazy.map(\.value)
+      }
+    case let .propertyAccess(value: value, keyPath: keyPath):
+      [value, keyPath]
+    }
+  }
+}
+
+extension Expression: Codable {}
+extension Expression.Kind: Codable {}
+extension Expression.Kind.FunctionCallArgument: Codable {}
+
+extension Expression: CustomStringConvertible, CustomDebugStringConvertible {
+  /// Initialize an instance of this type containing the specified source code.
+  ///
+  /// - Parameters:
+  ///   - sourceCode: The source code of the expression being described.
+  ///
+  /// To get the string value of an instance of ``Expression``, pass it to
+  /// `String.init(describing:)`.
+  ///
+  /// This initializer does not attempt to parse `sourceCode`.
+  public init(_ sourceCode: String) {
```

Comment:
If we make the type SPI, it's moot and I'm not worried. If it remains API, I do think it's useful but since we're not doing any parsing… perhaps minimally so. We could make it non-public if you feel strongly.

Reply:
Can we leave it public but make it `@_spi(ExperimentalSourceCodeCapturing)`?

Reply:
Done.